components/encryption/src/manager/mod.rs

```diff
@@ -98,75 +108,85 @@ impl Dicts {
+        let file_bytes = file_dict.write_to_bytes()?;
         // File dict is saved in plaintext.
+        let _lock = self.file_lock.lock().unwrap();
         file.write(&file_bytes, &PlaintextBackend::default())?;
 
         ENCRYPTION_FILE_SIZE_GAUGE
             .with_label_values(&["file_dictionary"])
             .set(file_bytes.len() as _);
-        ENCRYPTION_FILE_NUM_GAUGE.set(self.file_dict.files.len() as _);
+        ENCRYPTION_FILE_NUM_GAUGE.set(file_dict.files.len() as _);
 
         Ok(())
     }
 
-    fn get_key(&self, key_id: u64) -> Option<&DataKey> {
-        self.key_dict.keys.get(&key_id)
-    }
-
-    fn current_data_key(&self) -> (u64, &DataKey) {
+    fn current_data_key(&self) -> (u64, DataKey) {
+        let key_dict = self.key_dict.lock().unwrap();
+        let current_key_id = self.current_key_id.load(Ordering::Relaxed);
         (
-            self.key_dict.current_key_id,
-            self.key_dict
+            current_key_id,
+            key_dict
                 .keys
-                .get(&self.key_dict.current_key_id)
+                .get(&current_key_id)
+                .cloned()
                 .unwrap_or_else(|| {
                     panic!(
                         "current data key not found! Number of keys {}",
-                        self.key_dict.keys.len()
+                        key_dict.keys.len()
                     );
                 }),
         )
     }
 
-    fn get_file(&mut self, fname: &str) -> &FileInfo {
-        if self.file_dict.files.get(fname).is_none() {
-            // Return Plaintext if file not found
-            let mut file = FileInfo::default();
-            file.method = compat(EncryptionMethod::Plaintext);
-            self.file_dict.files.insert(fname.to_owned(), file);
+    fn get_file(&self, fname: &str) -> FileInfo {
+        let dict = self.file_dict.lock().unwrap();
+        let file_info = dict.files.get(fname);
+        match file_info {
+            None => {
+                // Return Plaintext if file not found
+                let mut file = FileInfo::default();
+                file.method = compat(EncryptionMethod::Plaintext);
+                file
+            }
+            Some(info) => info.clone(),
         }
-        self.file_dict.files.get(fname).unwrap()
     }
 
-    fn new_file(&mut self, fname: &str, method: EncryptionMethod) -> Result<&FileInfo> {
+    fn new_file(&self, fname: &str, method: EncryptionMethod) -> Result<FileInfo> {
         let iv = Iv::new_ctr();
         let mut file = FileInfo::default();
         file.iv = iv.as_slice().to_vec();
-        file.key_id = self.key_dict.current_key_id;
+        file.key_id = self.current_key_id.load(Ordering::Relaxed);
         file.method = compat(method);
-        self.file_dict.files.insert(fname.to_owned(), file);
+        let mut file_dict = self.file_dict.lock().unwrap();
+        file_dict.files.insert(fname.to_owned(), file.clone());
+        drop(file_dict);
```

Comment:
can we use scoping instead of drop? e.g.
```
{
    let mut file_dict = self.file_dict.lock().unwrap();
    file_dict.files.insert(fname.to_owned(), file.clone());
}
```